### PR TITLE
This enables every single parameter in a group of parameters to be specified.

### LINKED ISSUE #23867: Add ability to specify lower and upper bounds and intitial conditions for every parameter
## Reason
Lower and upper bounds and initial conditions can only be specified for each group of parameters being optimized.  It would be useful if these could be specified for every parameter in the group.  This is discussed here:
https://github.com/idaholab/moose/discussions/23772#discussioncomment-5426926
## Design
Make lower and upper bounds and initial conditions a vector of vectors, one vector of parameters for each parameter group being optimized.   These vector of vectors will need to be flattened and properly indexed when used by TAO.

## Impact
It will be useful to apply a range of different parameters to lower and upper bounds and initial conditions.

### PR DESCRIPTION
**Pull request by lynnmunday**
closes #23867

### CODE REVIEW

File: modules/optimization/src/optimizationreporters/OptimizationReporterBase.C [hunk 1]
```diff
@@ -110,27 +96,11 @@ OptimizationReporterBase::updateParameters(const libMesh::PetscVector<Number> &
 Real
 OptimizationReporterBase::getLowerBound(dof_id_type i) const
 {
-  return _lower_bounds.empty() ? std::numeric_limits<Real>::lowest()
-                               : _lower_bounds[getParameterIndex(i)];
+  return _lower_bounds[i];
```
Comment (lynnmunday) [inline modules/optimization/src/optimizationreporters/OptimizationReporterBase.C:99]: I started filling these completely so they don't need to be indexed into.  There are also always filled so there is no longer a need to check if bounds exist.

**Comment by loganharbour** [comment]
I take it you resolved your `moose` function issue? 😆

**Comment by lynnmunday** [comment]
@loganharbour I don't use for debugging anymore!

File: modules/optimization/src/optimizationreporters/ParameterMeshOptimization.C [hunk 2]
```diff
@@ -64,7 +67,10 @@ ParameterMeshOptimization::ParameterMeshOptimization(const InputParameters & par
   const auto & initial_condition = getParam<std::vector<Real>>("initial_condition");
   const auto & ntimes = getParam<unsigned int>("num_parameter_times");
 
-  // Size checks
+  std::vector<Real> constant_lower_bounds(getParam<std::vector<Real>>("lower_bounds"));
+  std::vector<Real> constant_upper_bounds(getParam<std::vector<Real>>("upper_bounds"));
```
Comment (zachmprince) [inline modules/optimization/src/optimizationreporters/ParameterMeshOptimization.C:71]:
```suggestion
  const std::vector<Real> & constant_lower_bounds = getParam<std::vector<Real>>("lower_bounds");
  const std::vector<Real> & constant_upper_bounds = getParam<std::vector<Real>>("upper_bounds");
```

**Comment by lynnmunday** [comment]
@zachmprince will you review?

File: modules/optimization/src/optimizationreporters/ParameterMeshOptimization.C [hunk 3]
```diff
@@ -96,5 +106,22 @@ ParameterMeshOptimization::ParameterMeshOptimization(const InputParameters & par
...
+                         vector_of_group_lower_bound.begin(),
+                         vector_of_group_lower_bound.end());
+
+    std::vector<Real> vector_of_group_upper_bound(_nvalues[i],
+                                                  constant_upper_bounds.empty()
+                                                      ? std::numeric_limits<Real>::max()
+                                                      : constant_upper_bounds[i]);
+    _upper_bounds.insert(_upper_bounds.end(),
+                         vector_of_group_upper_bound.begin(),
+                         vector_of_group_upper_bound.end());
```
Comment (zachmprince) [inline modules/optimization/src/optimizationreporters/ParameterMeshOptimization.C:125]:
```suggestion

    // assign lower and upper bounds
    _lower_bounds.resize(_lower_bounds.size() + _nvalues[i],
                                                       constant_lower_bounds.empty()
                                                      ? std::numeric_limits<Real>::lowest()
                                                      : constant_lower_bounds[i]);
    _upper_bounds.resize(_upper_bounds.size() + _nvalues[i],
                                                       constant_upper_bounds.empty()
                                                      ? std::numeric_limits<Real>::max()
                                                      : constant_upper_bounds[i]);
```

**Review by zachmprince — CHANGES_REQUESTED** [review]

File: modules/optimization/src/optimizationreporters/ParameterMeshOptimization.C (hunk 2 above)
Comment (lynnmunday) [reply modules/optimization/src/optimizationreporters/ParameterMeshOptimization.C:71]: why the reference?

File: modules/optimization/src/optimizationreporters/OptimizationReporter.C [hunk 4]
```diff
@@ -35,24 +45,98 @@ OptimizationReporter::OptimizationReporter(const InputParameters & parameters)
...
-    paramError("initial_condition",
-               "Initial condition must be same length as the total number of parameter values.");
-
-  unsigned int v = 0;
+    paramError(
+        "num_parameters",
+        "There should be a number in \'num_parameters\' for each name in \'parameter_names\'.");
+
+  std::vector<std::vector<Real>> initial_conditions(fillInitialConditions());
+  fillBounds();
```
[the remark this answers is not in the document]
Comment (lynnmunday) [reply modules/optimization/src/optimizationreporters/OptimizationReporter.C]: I switched to this

File: modules/optimization/src/optimizationreporters/ParameterMeshOptimization.C (hunk 2 above)
Comment (lynnmunday) [reply modules/optimization/src/optimizationreporters/ParameterMeshOptimization.C]: fixed

File: modules/optimization/src/optimizationreporters/OptimizationReporter.C [hunk 5]
```diff
@@ -35,24 +45,98 @@ OptimizationReporter::OptimizationReporter(const InputParameters & parameters)
...
   {
-    // initial conditions are different between this and ParameterMeshOptimization reporter
-    // these ICs are for every single parameter.  ParameterMeshOptimization defines a single
-    // condition for each group of parameters
-    _parameters[i]->assign(initial_condition.begin() + v,
-                           initial_condition.begin() + v + _nvalues[i]);
+    _parameters[i]->assign(initial_conditions[i].begin(), initial_conditions[i].end());
     _gradients[i]->resize(_nvalues[i]);
-    v += _nvalues[i];
   }
```
[the remark this answers is not in the document]
Comment (lynnmunday) [reply modules/optimization/src/optimizationreporters/OptimizationReporter.C]: fixed

File: modules/optimization/src/optimizationreporters/OptimizationReporter.C [hunk 6]
```diff
@@ -35,24 +45,98 @@ OptimizationReporter::OptimizationReporter(const InputParameters & parameters)
...
+      {
+        paramError(type,
+                   "When ",
+                   type,
+                   " are given in input file, there must either be a single value per parameter "
+                   "group or a value for every parameter in the group.");
+      }
+    }
+  }
+  return parsedData;
```
[the remark this answers is not in the document]
Comment (lynnmunday) [reply modules/optimization/src/optimizationreporters/OptimizationReporter.C]: fixed

**Comment by lynnmunday** [comment]
@zachmprince I made the changes you suggested, will you review again.

File: modules/optimization/src/optimizationreporters/OptimizationReporter.C [hunk 7]
```diff
@@ -35,24 +48,76 @@ OptimizationReporter::OptimizationReporter(const InputParameters & parameters)
...
+        parsedData[i] = params_group_constant;
+      }
+      else if ((parsedData[i].size() != 1) && (parsedData[i].size() != _nvalues[i]))
+      {
+        paramError(type,
+                   "When ",
+                   type,
+                   " are given in input file, there must either be a single value per parameter "
+                   "group or a value for every parameter in the group.");
+      }
```
Comment (zachmprince) [inline modules/optimization/src/optimizationreporters/OptimizationReporter.C:103]:
```suggestion
      if (parsedData[i].size() == 1)
        parsedData[i].assign(_nvalues[i], parsedData[i][0])
      else if (parsedData[i].size() != _nvalues[i])
        paramError(type,
                   "When ",
                   type,
                   " are given in input file, there must either be a single value per parameter "
                   "group or a value for every parameter in the group.");
```

File: modules/optimization/src/optimizationreporters/OptimizationReporter.C [hunk 8]
```diff
@@ -35,24 +48,76 @@ OptimizationReporter::OptimizationReporter(const InputParameters & parameters)
...
+  }
+
+  if (parsedData.empty())
+  {
+    // fill with default values
+    for (auto & params_per_group : _nvalues)
+    {
+      std::vector<Real> param_group_defaults(params_per_group, default_value);
+      parsedData.push_back(param_group_defaults);
+    }
```
Comment (zachmprince) [inline modules/optimization/src/optimizationreporters/OptimizationReporter.C:114]:
```suggestion
    for (const auto & params_per_group : _nvalues)
      parsed_data.emplace_back(params_per_group, default_value);
```

**Review by zachmprince — CHANGES_REQUESTED** [review]
Just making things a little more concise. You can take or leave it.

File: modules/optimization/src/optimizationreporters/OptimizationReporter.C (hunk 7 above)
Comment (lynnmunday) [reply modules/optimization/src/optimizationreporters/OptimizationReporter.C:103]: finished

File: modules/optimization/src/optimizationreporters/OptimizationReporter.C (hunk 8 above)
Comment (lynnmunday) [reply modules/optimization/src/optimizationreporters/OptimizationReporter.C:114]: done

File: modules/optimization/src/optimizationreporters/OptimizationReporter.C [hunk 9]
```diff
@@ -35,24 +48,76 @@ OptimizationReporter::OptimizationReporter(const InputParameters & parameters)
...
   }
+
+  // flatten into single vector
+  std::vector<Real> flattened_data;
+  for (auto & vec : parsedData)
```
[the remark this answers is not in the document]
Comment (lynnmunday) [reply modules/optimization/src/optimizationreporters/OptimizationReporter.C:119]: done

**Review by zachmprince — DISMISSED** [review]

**Review by zachmprince — APPROVED** [review]

**Pull request merged**